packages/babel-plugin-transform-react-jsx-self/src/index.js

```diff
@@ -16,11 +16,78 @@ import { types as t } from "@babel/core";
 
 const TRACE_ID = "__self";
 
+/**
+ * Returns whether the given path contains a `super()` call.
+ */
+function containsSuperCall(sourcePath) {
+  // We don't want to go into inner classes as their super() calls aren't relevant for us.
+  if (sourcePath.isClassDeclaration() || sourcePath.isClassExpression()) {
+    return false;
+  }
+  let exists = false;
+  sourcePath.traverse({
+    // Again, we don't go into inner classes.
+    ClassDeclaration(path) {
+      path.skip();
+    },
+    ClassExpression(path) {
+      path.skip();
+    },
+    // Find a `Super` that is used in a call expression.
+    Super(path) {
+      if (path.parentPath.isCallExpression()) {
+        exists = true;
+        path.stop();
+      }
+    },
+  });
+  return exists;
+}
+
+/**
+ * Returns whether it is allowed to use `this` at given path.
+ */
+function canUseThis(path) {
+  const statement = path.getStatementParent();
+  // This specifically skips arrow functions as they do not rewrite `this`.
+  const parentMethodOrFunction = statement.findParent(path => {
+    return (
+      path.isFunctionDeclaration() ||
+      path.isFunctionExpression() ||
+      path.isClassMethod()
+    );
+  });
+  if (parentMethodOrFunction === null) {
+    // We are not in a method or function. It is fine to use `this`.
+    return true;
+  }
+  if (!parentMethodOrFunction.isMethod()) {
+    // If the closest parent is a regular function, `this` will be rebind, therefore it is fine to use `this`.
+    return true;
+  }
+  // Current node is within a method, so we need to check if the method is a constructor.
+  if (parentMethodOrFunction.node.kind !== "constructor") {
+    // We are not in a constructor, therefore it is always fine to use `this`.
+    return true;
+  }
+  // Now we are in a constructor. We need to check if there is a `super()` call following the current node.
+  if (containsSuperCall(statement)) {
+    return false;
+  }
+  return !statement
```

Comment:
I think `containsSuperCall` should be invoked on `parentMethodOrFunction` so we don't have to duplicate logic for siblings.

Reply:
Since we also have to make sure the `super()` call is located after the jsx tag, I am not sure if this is viable.

Reply:
I think we can fallback to not adding `this` for `constructor` including `Super` call, because the AST order does not necessarily yield the execution order, for example:

```js
if (random() > 0.5) {
  <jsx/>
} else {
  super()
}
```

Reply:
Got it. Thanks.

Reply:
Done.